Initial implementation Feature: Remote Buzzer Server

jacques42 opened the pull request:
Initial implementation **Feature: Remote Buzzer Server** based on io sockets

Fix https://github.com/andi34/photobooth/issues/93

jacques42 committed:
Initial implementation Feature: Remote Buzzer Server

Change-Id: I8f723c165175a4ed99ced07cba11596ca0f1d006

andi34 commented:
Thanks a lot! :)

Could you please take a look at the lint test fail?
https://github.com/andi34/photobooth/runs/968714117?check_suite_focus=true

jacques42 committed:
robustness and improvements

load photobooth config to remotebuzzer_server
adjusted install script and dependencies
eslint sanity and clean-up

Change-Id: Ic9494fa1324855e5c93a180faaa9a8edc1a0bd92

jacques42 commented:
I've done some improvement work on the code and also the eslint sanity. However the Lint check still failes for reasons and on files that I have not touched. Not sure but I assume I have to rebase on the latest 2.8 branch and build on top. 

I am going to be away for the coming two weeks, so please bear with me until early September for to be able to come back to this.

andi34 commented:
Thanks and take your time :)
Maybe lint fails because of your changes to .eslintrc.json?

jacques42 commented:
I definitely appreciate if you would take a look. The [eslint rule that throws the errors](https://eslint.org/docs/rules/no-unused-vars) is `no-unused-vars` and is derived from the base rule set (`"extends": "eslint:recommended"`), which has been in the config file since.

```
/home/runner/work/photobooth/photobooth/src/js/adminshortcut.js
  8:10  error  'adminsettings' is defined but never used  no-unused-vars
/home/runner/work/photobooth/photobooth/src/js/chromakeying.js
  82:10  error  'setBackgroundImage' is defined but never used  no-unused-vars
/home/runner/work/photobooth/photobooth/src/js/photoinit.js
  3:10  error  'initPhotoSwipeFromDOM' is defined but never used  no-unused-vars
```
Changes I've done in the commits of this PR have not touched those three files, so I have no idea why this rule tiggers now. Of course I can remove the variables from the code to fix it. But in order to be better safe than sorry, I appreciate you looking at it as surely you have a better perspective on what's going on there.

The changes I've done to `.eslintrc.json` where for two reasons:

1. Error handler implementation
I am using a process error handler for `uncaughtExceptions`. If that triggers, the code does some clean-up and logging, and then the process must exit by all means, as something went wrong in a non-recoverable way. So I use the function `process.exit()` in that exception handler. The use of `process.exit()` triggered eslint rule `no-process-exit`. In the [rule documentation ](https://eslint.org/docs/rules/no-process-exit) it is marked as depreciated since eslint v7, with a recommendation to include the plugin `eslint-plugin-node`. So I did include this plug-in, added the `plugin:node/recommended` rule set and then turned off the depreciated rule `no-process-exit` in the `.eslintrc.json` configuration file.

2. Loading of Photobooth configuration 
At the start of the process, it reads the Photobooth current configuration. For that I decided to use a syncronous (blocking) function call `execSync`. Reason being is I'd like for this operation to complete before any further processing happens - simply to ensure operation completes first so that the Photobooth configuration data is available onwards. That synchronous function call triggered eslint rule `no-sync`. Similarly, the [rule documentation ](https://eslint.org/docs/rules/no-sync) says this rule is depreciated as of ESlint v7 and recommends to use the `eslint-plugin-node`. So I did accordingly, including to turn of rule `no-sync` in the `.eslintrc.json` configuration file.

Currently I can't see a good reason why that changes to `.eslintrc.json` could have caused to throw above three errors as new. But I don't know what I don't know. So I definitely appreciate your help looking into this. 

And of course, if there are better ways / better code to implement my items 1) and 2) I am super eager to learn. Please let me know. 

Talk to you in September !

jacques42 committed:
eslint sanity

Change-Id: I811143f022e3fc78da8c3a89156ff77f29636cce

andi34 commented:
Hey and thanks for taking a look at the lint errors.
Don't worry about the dev branch here being ahead of yours, we can rebase your commits once everything is setup.
There's a reason for "Revert "fix (slideshow): exit slideshow on close""?

jacques42 commented:
Removing it was a mistake, I added it back in.

jacques42 committed:
manual and faq updates

Change-Id: Ie4526675eb6cc4ec10608d35af2af4387bf18584

andi34 committed:
Adjust eslint config to match Prettier setup

Change-Id: I33f38a8058e6388bd03a74d7cd58cf931311d0f3

andi34 committed:
yarn.lock: regenerate

Change-Id: Ibf46c8bf4cf67e248241313e12c40c6b43875b65

jacques42 committed:
Updated install-raspbian.sh and scripts/pack-build.js for remote buzzer feature

jacques42 committed:
remote buzzer: added one more depedency to pack-build.js

andi34 commented on install-raspbian.sh:
```diff
@@ -269,6 +269,16 @@ EOF
 
 fi
 
+info "### Enable Nodejs GPIO access - please reboot"
```
Maybe optional via request?

jacques42 replied:
The new overlay in /boot/config.txt will only activate at reboot. I did not find any way to activate this without reboot. And without this overlay, there is stability issues it seems.

Of course the reboot at this point is optional and not using the remote buzzer feature, the reboot is not required at all. 

I would not add a prompt here as right at the end of the script, anyways the user is prompted for a possible reboot.

jacques42 committed:
Be more clear for reboot required and added npm package to install as for a new buster installation it was not there.

andi34 committed:
compatibility (config): disable Remote Buzzer Server on Windows

Change-Id: I64c4ae80b93ffe312b6ebeaf822d83509208ce8a

andi34 committed:
update-booth: add nodejs

Change-Id: I0d10f10b57345357b32c8307f08020b94bdbb0c7
